### ISSUE TITLE
import/no-cycle throwing `No "exports" main defined on package`

### ISSUE THREAD

**Issue by MattIPv4** [import-js/eslint-plugin-import#2794]
When running eslint across a repo, I am encountering the following when importing one specific package:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in [project]/node_modules/[package]/package.json
Occurred while linting [project]/[file]/index.js:8
Rule: "import/no-cycle"
    at new NodeError (node:internal/errors:399:5)
    at exportsNotFound (node:internal/modules/esm/resolve:361:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:641:13)
    at resolveExports (node:internal/modules/cjs/loader:565:36)
    at Module._findPath (node:internal/modules/cjs/loader:634:31)
    at findModulePath ([project]/node_modules/eslint-import-resolver-alias/index.js:99:27)
    at exports.resolve ([project]/node_modules/eslint-import-resolver-alias/index.js:75:10)
    at v2 ([project]/node_modules/eslint-module-utils/resolve.js:116:23)
    at withResolver ([project]/node_modules/eslint-module-utils/resolve.js:121:14)
    at fullResolve ([project]/node_modules/eslint-module-utils/resolve.js:138:22)
```

The `package.json` of this dependency contains the following:

```
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./heroes": {
      "import": "./dist/heroes/index.js",
      "types": "./dist/heroes/index.d.ts"
    },
    "./icons": {
      "import": "./dist/icons/index.js",
      "types": "./dist/icons/index.d.ts"
    }
  }
```

What exports is import/no-cycle expecting here that this package is missing?

**Comment by ljharb** [comment]
That error is coming from node itself, and eslint-import-resolver-alias. I'm not sure what's causing it, but you'l need to report it there.

**Comment by MattIPv4** [comment]
Hm, interesting. This package seems to import fine in an ESM environment. I'll do some source diving to try and figure out how Node is loading this that results in it tripping up (on what seems to be a valid package). Ty!

**Comment by ljharb** [comment]
I agree the package seems valid.

**Comment by pyryk** [comment]
I encountered a similar problem when running eslint-import rules with `eslint-import-resolver-alias` enabled on an app that has msw@2.0.9 as a dependency. Not sure if `eslint-import-resolver-alias` is required for this issue to reproduce or not. At least in that case, the problem seems to originate from the fact that for `msw/browser`, there is no conditional export definition for node:

```json
"exports": {
    ".": {
      "types": "./lib/core/index.d.ts",
      "require": "./lib/core/index.js",
      "import": "./lib/core/index.mjs",
      "default": "./lib/core/index.js"
    },
    "./browser": {
      "node": null,
      "types": "./lib/browser/index.d.ts",
      "require": "./lib/browser/index.js",
      "import": "./lib/browser/index.mjs",
      "default": "./lib/browser/index.js"
    },
    ...
}
```

If I'm understanding everything right, the package.json snippet in the original issue description does not contain any node-compatible entry (`node` or `default`) either. Therefore, these issues could have the same root cause.

I'm assuming the node resolver always looks for the node version of the package, even in cases when the code is only ever supposed to be run in the browser.

In my case, I was able to work around the problem with either of these:

1. replace `"node": null` with `"node": "./lib/browser/index.js"` in the dependency package.json (_this is a hack at best and probably not really recommended_)
2. Add the following alias mapping to the `import/resolver` settings (I'm assuming this requires `eslint-import-resolver-alias` which has not been updated in several years, but seems to work OK for my use case):

```json
"settings": {
   ...
    "import/resolver": {
      ...
      "alias": {
        "map": [
          ["msw/browser", "node_modules/msw/lib/core/index.js"]
        ]
      }
    }
  }
```
